presto-parquet/src/main/java/com/facebook/presto/parquet/predicate/TupleDomainParquetPredicate.java

```diff
@@ -162,85 +192,74 @@ public static Domain getDomain(Type type, long rowCount, Statistics<?> statistic
             }
+            return Domain.create(ValueSet.ofRanges(ranges), hasNullValue);
+        }
 
-            if (doubleStatistics.genericGetMin().isNaN() || doubleStatistics.genericGetMax().isNaN()) {
-                return Domain.create(ValueSet.all(type), hasNullValue);
-            }
+        if (type.equals(DOUBLE)) {
+            List<Range> ranges = new ArrayList<>();
+            for (int i = 0; i < minimums.size(); i++) {
+                Double min = (Double) minimums.get(i);
+                Double max = (Double) maximums.get(i);
+                if (min.isNaN() || max.isNaN()) {
+                    return Domain.create(ValueSet.all(type), hasNullValue);
+                }
 
-            ParquetDoubleStatistics parquetDoubleStatistics = new ParquetDoubleStatistics(doubleStatistics.genericGetMin(), doubleStatistics.genericGetMax());
-            return createDomain(type, hasNullValue, parquetDoubleStatistics);
+                ranges.add(Range.range(type, min, true, max, true));
+            }
+            return Domain.create(ValueSet.ofRanges(ranges), hasNullValue);
         }
 
-        if (isVarcharType(type) && statistics instanceof BinaryStatistics) {
-            BinaryStatistics binaryStatistics = (BinaryStatistics) statistics;
-            Slice minSlice = Slices.wrappedBuffer(binaryStatistics.genericGetMin().getBytes());
-            Slice maxSlice = Slices.wrappedBuffer(binaryStatistics.genericGetMax().getBytes());
-            if (minSlice.compareTo(maxSlice) > 0) {
-                failWithCorruptionException(failOnCorruptedParquetStatistics, column, id, binaryStatistics);
-                return Domain.create(ValueSet.all(type), hasNullValue);
+        if (isVarcharType(type)) {
+            List<Range> ranges = new ArrayList<>();
+            for (int i = 0; i < minimums.size(); i++) {
+                Slice min = Slices.wrappedBuffer(((Binary) minimums.get(i)).toByteBuffer());
+                Slice max = Slices.wrappedBuffer(((Binary) maximums.get(i)).toByteBuffer());
+                ranges.add(Range.range(type, min, true, max, true));
             }
-            ParquetStringStatistics parquetStringStatistics = new ParquetStringStatistics(minSlice, maxSlice);
-            return createDomain(type, hasNullValue, parquetStringStatistics);
+            return Domain.create(ValueSet.ofRanges(ranges), hasNullValue);
         }
 
-        if (type.equals(DATE) && statistics instanceof IntStatistics) {
-            IntStatistics intStatistics = (IntStatistics) statistics;
-            if (intStatistics.genericGetMin() > intStatistics.genericGetMax()) {
-                failWithCorruptionException(failOnCorruptedParquetStatistics, column, id, intStatistics);
-                return Domain.create(ValueSet.all(type), hasNullValue);
+        if (type.equals(DATE)) {
+            List<Range> ranges = new ArrayList<>();
```

Comment:
nit: `List<Range> ranges = new ArrayList<>();` and `Domain.create(ValueSet.ofRanges(ranges), hasNullValue);` is common in four places. Declare the array outside these blocks and use it create the ValueSet at the end of the function?

Reply:
yep, `List<Range> ranges = new ArrayList<>()` could move outside